codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -285,193 +321,502 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
-
-        ChannelFuture future = group.register(channel);
-        // Handle any errors that occurred on the local thread while registering. Even though
-        // failures can happen after this point, they will be handled by the channel by closing the
-        // channel.
-        if (future.cause() != null) {
-            if (channel.isRegistered()) {
-                channel.close();
-            } else {
-                channel.unsafe().closeForcibly();
-            }
-        }
-        return future;
-    }
+    private static final class DefaultHttp2StreamChannel extends AbstractChannel implements Http2StreamChannel {
 
-    @SuppressWarnings("unchecked")
-    private static void initOpts(Channel channel, Map<ChannelOption<?>, Object> opts) {
-        if (opts != null) {
-            for (Entry<ChannelOption<?>, Object> e: opts.entrySet()) {
-                try {
-                    if (!channel.config().setOption((ChannelOption<Object>) e.getKey(), e.getValue())) {
-                        logger.warn("Unknown channel option: " + e);
-                    }
-                } catch (Throwable t) {
-                    logger.warn("Failed to set a channel option: " + channel, t);
-                }
-            }
-        }
-    }
+        private final ChannelHandlerContext ctx;
+        private final ChannelFutureListener firstFrameWriteListener;
 
-    @SuppressWarnings("unchecked")
-    private static void initAttrs(Channel channel, Map<AttributeKey<?>, Object> attrs) {
-        if (attrs != null) {
-            for (Entry<AttributeKey<?>, Object> e: attrs.entrySet()) {
-                channel.attr((AttributeKey<Object>) e.getKey()).set(e.getValue());
-            }
-        }
-    }
+        /** {@code true} after the first HEADERS frame has been written **/
+        private boolean firstFrameWritten;
 
-    final class Http2StreamChannel extends AbstractHttp2StreamChannel implements ChannelFutureListener {
-        boolean onStreamClosedFired;
+        /** {@code true} if a close without an error was initiated **/
+        private boolean streamClosedWithoutError;
+
+        /** {@code true} if stream is in {@link Http2MultiplexCodec#channelsToFireChildReadComplete}. **/
+        boolean inStreamsToFireChildReadComplete;
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<DefaultHttp2StreamChannel> OUTBOUND_FLOW_CONTROL_WINDOW_UPDATER;
 
         /**
-         * {@code true} if stream is in {@link Http2MultiplexCodec#channelsToFireChildReadComplete}.
+         * Used by subclasses to queue a close channel within the read queue. When read, it will close
+         * the channel (using Unsafe) instead of notifying handlers of the message with {@code
+         * channelRead()}. Additional inbound messages must not arrive after this one.
          */
-        boolean inStreamsToFireChildReadComplete;
+        protected static final Object CLOSE_MESSAGE = new Object();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this "protected" modifier. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2156)